---
title: "Cleaning Pieris FASTA"
author: "Katie McLaughlin & James R. Walters"
output: 
  html_document: 
    highlight: pygments
    theme: readable
    toc: true
    toc_float: true
    toc_collapsed: false

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries & data
```{r}
suppressMessages(library(Biostrings))

```


Protein predictions from genome
```{r}
pierisGenome <- readAAStringSet(file = "../../PierisGenomeData/Pieris_rapae_proteins.fa")
```

INTERMEDIATE STEP - Used Linux: in the gff file, used the grep command in conjunction with cut to write out to file the gene names and gene IDs

`grep "\tgene\t" Pieris_rapae_pierap_codingGene_final.gff | cut -f9 > geneID.txt`

Now clean up this intermediate file to give Transcript ID, Gene Name, and GeneID.

```{r}

##Generating the Table 
geneinfo <- read.table(file = "../../PierisGenomeData/geneID.txt", sep =";", 
                       header = FALSE, quote = "", as.is = T) #15047 gene IDs

```

## Clean GeneIDs

```{r}
#Update column names in the table
names(geneinfo)[1] <- "GeneID"
names(geneinfo)[2] <- "GeneName"
protNames <- names(pierisGenome)

#Remove "ID=" and "Name=" 
geneinfo$GeneID <- gsub("ID=", "", geneinfo$GeneID)
geneinfo$GeneName <- gsub("Name=", "", geneinfo$GeneName)

#Use the `match()` function to subset geneinfo$GeneID in a way that is consistent with the ordering and identity of transcript IDs on AAstringset read from fasta.
trxList <- strsplit(names(pierisGenome), split = " gene=")  #Parse fasta header from AAStringSet into list of unique IDs

prot.mtx <- do.call(what = rbind, args = trxList) #Turn this list into a matrix, to access only trxID or GeneName directly 
names(prot.mtx)[1] <- "TranscriptID"
names(prot.mtx)[2] <- "GeneName"
prot.geneID <- geneinfo$GeneID[match(prot.mtx[,2], table = geneinfo$GeneName )] #Use match to get GeneID values corresponding to GeneNames in the fasta ordering

#Combine into one table
trxID.geneName.geneID <- cbind(prot.mtx, prot.geneID)
#Fix names in table
colnames(trxID.geneName.geneID) <- c("TrxID", "GeneName", "GeneID")

write.csv(x = trxID.geneName.geneID[order(trxID.geneName.geneID[,1]), ], file = "../Results/TrxID.GeneName.GeneID.csv", row.names = F, quote = T)

```

# Update FASTA headers

Update headers in FASTA entry to contain GeneID

Paste together to give the new IDs for the AAStringSet:

`newNames <- paste(prot.mtx[,1], prot.geneID, prot.mtx[,2])` 
```{r}
newNamesKeys <- paste0(prot.mtx[,1], " geneID=", prot.geneID, " genename=", prot.mtx[,2])  
names(pierisGenome) <- newNamesKeys
## length(names(pierisGenome)) 

#EXTRA: Clean the sequences (remove ".")
tmpchar <- as.character(pierisGenome)
tmpchar <- gsub(pattern = ".", x = tmpchar, fixed = T, replacement = "")
pierisGenome <- AAStringSet(tmpchar)

  #Write out FASTA file
writeXStringSet(x = pierisGenome, filepath = "../Results/CleanedProteomeFASTAs/PierisRenamedProteins.fas") 

```


# Remove redundant sequences

Many proteins are identical.  Need to remove redundant proteins, but keep all unique isoforms for each gene.
```{r}
geneAAsetsList <- split(pierisGenome, f = prot.mtx[,2]) 
#Then use lapply and unique
geneAAsetsListuniq <- lapply(geneAAsetsList, unique) #Standard list output 
## geneAAsetsListuniq <- unlist(geneAAsetsListuniq)
## str(geneAAsetsListuniq)
## length(geneAAsetsListuniq) #Sanity check

#Fix the data structure
geneAAsetsListuniq <- AAStringSetList(geneAAsetsListuniq)
names(geneAAsetsListuniq) <- NA
uniqAAstringset <- unlist(geneAAsetsListuniq) #To get back to StringSet
## names(uniqAAstringset[1:3])

#Write out FASTA file
writeXStringSet(x = uniqAAstringset, 
                filepath = "../Results/CleanedProteomeFASTAs/PierisNonRedundant.fas") 

```

# Get longest CDS

Need to select a single protein per gene for comparative/orthology analyses.  We will choose the longest one.

```{r}
indexSeq <- which.max(width(geneAAsetsListuniq[[2]])) #Index that reflects the longest transcript 


#Create a function 
getLongest <- function(x) {
  indexSeq <- which.max(width(x))
  longestSSObject <- x[indexSeq]
  return(longestSSObject)
}

#Put defined function into lapply 
longestSeqList <- lapply(geneAAsetsList, FUN = getLongest) 

#Fix the data structure
longestCDS <- AAStringSetList(longestSeqList) #Convert from standard list to StringSetList
names(longestCDS) <- NA #Remove list names
longestCDS <- unlist(longestCDS)

#Write out FASTA file
writeXStringSet(x = longestCDS, filepath = "../Results/CleanedProteomeFASTAs/PierisLongestCDS.fas") 

```
## Reformat to use GeneID is primary identifier

Subsequent annotation efforts would be streamlined for merging with other sequencees if the sequence identifier were the gene and not the transcript ID.

```{r}
longestIDs <- names(longestCDS)
newIDs <- gsub(pattern = "(PIERAPT\\d+) geneID=(PIERAPG\\d+) (genename=.+)", 
               replacement = "\\2 trxID=\\1 \\3", 
               x = longestIDs , perl = T)

# duplicate object to avoid any future problmes
# longestCDS is used again.
longestCDSgeneID <- longestCDS 
names(longestCDSgeneID) <- newIDs

#Write out FASTA file
writeXStringSet(x = longestCDSgeneID, filepath = "../Results/CleanedProteomeFASTAs/PierisLongestCDS_geneID.fas") 

```


# Visualize redundancies

```{r}
pdf(file = "../Results/GenomePlots.pdf")
#The plot below demonstrates per gene the redundant and unique versions of transcript: 
## geneID <- strsplit(names(pierisGenome), split = " gene=") 
## gene.mtx <- do.call(what = rbind, args = geneID)
## colnames(gene.mtx)[1] <- "TrxID"
## colnames(gene.mtx)[2] <- "GeneName"
## geneAAsetsList <- split(pierisGenome, f = gene.mtx[ ,2]) 
lengthSS <- sapply(geneAAsetsList, length)

## geneAAsetsListuniq <- lapply(geneAAsetsList, unique)
lengthSSUnique <- sapply(geneAAsetsListuniq, length) #Length = count of transcripts

plot(x = lengthSS, y = lengthSSUnique, pch = 19, xlab = "Original Count", 
     ylab = "Unique Transcripts", col = rgb(0,0,0,0.1))

#This barplot shows the difference (in counts) between the original and unique counts of the genome.
diffLengths <- lengthSS - lengthSSUnique
diffLengthsTable <- table(diffLengths) #Tabulate the counts

barplot(diffLengthsTable, xlab = "Count Length", ylab = "Frequency",
        main = "Difference Between Original and Unique Transcript Length", ylim = c(0,12000), space = 0)

#This barplot shows the number of unique transcripts versus gene length. 
geneLength <- sapply(longestSeqList, FUN = width) 

plot(x = geneLength, y = lengthSSUnique, pch = 19, xlab = "Gene Length", 
     ylab = "Unique Transcripts", col = rgb(0,0,0,0.1))

dev.off()
```

